src/test/ui/intrinsics/is_const_eval.rs

```diff
@@ -0,0 +1,11 @@
+// run-pass
+
+#![feature(core_intrinsics)]
+use std::intrinsics::is_const_eval;
+
+fn main() {
+    const X: bool = unsafe { is_const_eval() };
```

Comment:
So AFAICT the only issue is that this currently works due to #61495 allowing all intrinsics to run in const-contexts (notice that `feature(const_fn)` is not enabled!).  

Even if we were to fix #61495, this would still "just work" behind the `const_fn` feature. That would allow it to "silently" be used in APIs exposed by libstd. I'm not sure how to fix that.

Reply:
If it is fixed as I imagine it, the const fn feature gate won't have an effect on it. It would only permit a whitelist of intrinsics and forbid all others outside unleash mode

Reply:
Agreed with @oli-obk. Why would just "just fork" with `const_fn`? That gate won't make non-const-fn callable, and all intrinsics (except for a whitelist) should be considered non-const-fn.

Reply:
I think the "why" issue is already resolved in the updated OP description and the new test that shows an use case that works with this, but wouldn't work without it.